File: backend/internal/finance/finance_test.go
package finance

import (
	"database/sql"
	"math"
	"testing"
	"time"

	"github.com/google/uuid"
)

// ─── Aggregation math: SpendingBreakdown share computation ────────────────────

// simulateSpendingBreakdown mirrors the share-computation logic from
// Store.SpendingBreakdown so we can test it without a DB.
func simulateSpendingBreakdown(rows []spendingRow) []CategoryTotal {
	var grandTotal float64
	for _, r := range rows {
		grandTotal += r.TotalAmount
	}

	out := make([]CategoryTotal, 0, len(rows))
	for _, r := range rows {
		ct := CategoryTotal{
			TotalAmount: r.TotalAmount,
			TxCount:     r.TxCount,
		}
		if r.CategoryID.Valid {
			ct.CategoryID = r.CategoryID.UUID
		}
		if r.CategoryName.Valid {
			ct.CategoryName = r.CategoryName.String
		} else {
			ct.CategoryName = "Uncategorized"
		}
		if r.Kind.Valid {
			ct.Kind = r.Kind.String
		} else {
			ct.Kind = "expense"
		}
		if grandTotal > 0 {
			ct.SharePercent = (r.TotalAmount / grandTotal) * 100
		}
		out = append(out, ct)
	}
	return out
}

func TestSpendingShareSumsToHundred(t *testing.T) {
	rows := []spendingRow{
		{
			CategoryID:   uuid.NullUUID{UUID: uuid.New(), Valid: true},
			CategoryName: nullString("Groceries"),
			Kind:         nullString("expense"),
			TotalAmount:  400,
			TxCount:      5,
		},
		{
			CategoryID:   uuid.NullUUID{UUID: uuid.New(), Valid: true},
			CategoryName: nullString("Transport"),
			Kind:         nullString("expense"),
			TotalAmount:  300,
			TxCount:      3,
		},
		{
			CategoryID:   uuid.NullUUID{UUID: uuid.New(), Valid: true},
			CategoryName: nullString("Utilities"),
			Kind:         nullString("expense"),
			TotalAmount:  300,
			TxCount:      2,
		},
	}
	totals := simulateSpendingBreakdown(rows)

	var sumShares float64
	for _, ct := range totals {
		sumShares += ct.SharePercent
	}
	if math.Abs(sumShares-100.0) > 0.001 {
		t.Errorf("shares sum = %.4f, want 100", sumShares)
	}
}

func TestSpendingShareCorrectValues(t *testing.T) {
	rows := []spendingRow{
		{
			CategoryID:   uuid.NullUUID{UUID: uuid.New(), Valid: true},
			CategoryName: nullString("Groceries"),
			Kind:         nullString("expense"),
			TotalAmount:  600,
			TxCount:      4,
		},
		{
			CategoryID:   uuid.NullUUID{UUID: uuid.New(), Valid: true},
			CategoryName: nullString("Transport"),
			Kind:         nullString("expense"),
			TotalAmount:  400,
			TxCount:      2,
		},
	}
	totals := simulateSpendingBreakdown(rows)

	if len(totals) != 2 {
		t.Fatalf("expected 2 categories, got %d", len(totals))
	}
	if math.Abs(totals[0].SharePercent-60.0) > 0.001 {
		t.Errorf("Groceries share = %.4f, want 60", totals[0].SharePercent)
	}
	if math.Abs(totals[1].SharePercent-40.0) > 0.001 {
		t.Errorf("Transport share = %.4f, want 40", totals[1].SharePercent)
	}
}

func TestSpendingZeroGrandTotal(t *testing.T) {
	// When there are no transactions, shares must all be 0 (no division by zero).
	rows := []spendingRow{
		{
			CategoryID:   uuid.NullUUID{UUID: uuid.New(), Valid: true},
			CategoryName: nullString("Groceries"),
			Kind:         nullString("expense"),
			TotalAmount:  0,
			TxCount:      0,
		},
	}
	totals := simulateSpendingBreakdown(rows)
	if totals[0].SharePercent != 0 {
		t.Errorf("share should be 0 when grand total is 0, got %.4f", totals[0].SharePercent)
	}
}

func TestSpendingUncategorizedBucket(t *testing.T) {
	// Rows without category should get name "Uncategorized".
	rows := []spendingRow{
		{
			CategoryID:   uuid.NullUUID{Valid: false},
			CategoryName: nullString(""),   // invalid (empty) → Uncategorized
			Kind:         nullString(""),   // invalid → "expense"
			TotalAmount:  200,
			TxCount:      1,
		},
	}
	// Override: blank CategoryName.Valid = false
	rows[0].CategoryName.Valid = false

	totals := simulateSpendingBreakdown(rows)
	if totals[0].CategoryName != "Uncategorized" {
		t.Errorf("CategoryName = %q, want Uncategorized", totals[0].CategoryName)
	}
	if totals[0].Kind != "expense" {
		t.Errorf("Kind = %q, want expense", totals[0].Kind)
	}
}

// ─── Net-worth math ───────────────────────────────────────────────────────────

// simulateNetWorth mirrors NetWorthNow arithmetic without any DB.
func simulateNetWorth(assets, holdings, liabs []valueRow, fxRates map[string]float64, baseCurrency string) NetWorthSnapshot {
	convert := func(amount float64, currency string) float64 {
		if currency == baseCurrency || currency == "" {
			return amount
		}
		if rate, ok := fxRates[currency]; ok && rate > 0 {
			return amount / rate
		}
		return amount
	}

	var totalAssets, totalHoldings, totalLiabs float64
	for _, a := range assets {
		totalAssets += convert(a.value, a.currency)
	}
	for _, h := range holdings {
		totalHoldings += convert(h.value, h.currency)
	}
	for _, l := range liabs {
		totalLiabs += convert(l.value, l.currency)
	}

	return NetWorthSnapshot{
		AsOf:          time.Now(),
		BaseCurrency:  baseCurrency,
		TotalAssets:   totalAssets,
		TotalHoldings: totalHoldings,
		TotalLiabs:    totalLiabs,
		NetWorth:      totalAssets + totalHoldings - totalLiabs,
	}
}

func TestNetWorthSimple(t *testing.T) {
	assets := []valueRow{{value: 500_000, currency: "ZAR"}}
	liabs := []valueRow{{value: 200_000, currency: "ZAR"}}
	snap := simulateNetWorth(assets, nil, liabs, nil, "ZAR")

	if snap.NetWorth != 300_000 {
		t.Errorf("net worth = %.2f, want 300000", snap.NetWorth)
	}
	if snap.TotalAssets != 500_000 {
		t.Errorf("total assets = %.2f, want 500000", snap.TotalAssets)
	}
	if snap.TotalLiabs != 200_000 {
		t.Errorf("total liabs = %.2f, want 200000", snap.TotalLiabs)
	}
}

func TestNetWorthWithHoldings(t *testing.T) {
	assets := []valueRow{{value: 100_000, currency: "ZAR"}}
	holdings := []valueRow{{value: 50_000, currency: "ZAR"}}
	liabs := []valueRow{{value: 30_000, currency: "ZAR"}}
	snap := simulateNetWorth(assets, holdings, liabs, nil, "ZAR")

	want := 100_000.0 + 50_000.0 - 30_000.0
	if snap.NetWorth != want {
		t.Errorf("net worth = %.2f, want %.2f", snap.NetWorth, want)
	}
}

func TestNetWorthFXNormalization(t *testing.T) {
	// 1 ZAR = 0.053 USD  → 1 USD = 18.87 ZAR
	// fxRates[quote] = rate where base = ZAR; so fxRates["USD"] = 0.053
	// To convert USD → ZAR: amount / rate = amount / 0.053
	fxRates := map[string]float64{"USD": 0.053}

	// Asset of 10 000 USD should convert to 10000 / 0.053 ≈ 188 679 ZAR.
	assets := []valueRow{{value: 10_000, currency: "USD"}}
	snap := simulateNetWorth(assets, nil, nil, fxRates, "ZAR")

	want := 10_000.0 / 0.053
	if math.Abs(snap.TotalAssets-want) > 1 {
		t.Errorf("FX-converted assets = %.2f, want %.2f", snap.TotalAssets, want)
	}
}

func TestNetWorthFXNoRate(t *testing.T) {
	// When no FX rate is available the amount is kept as-is (best-effort).
	fxRates := map[string]float64{} // empty
	assets := []valueRow{{value: 5_000, currency: "GBP"}}
	snap := simulateNetWorth(assets, nil, nil, fxRates, "ZAR")

	// Best-effort: keep original value.
	if snap.TotalAssets != 5_000 {
		t.Errorf("no-rate fallback: assets = %.2f, want 5000", snap.TotalAssets)
	}
}

func TestNetWorthNegative(t *testing.T) {
	// Liabilities larger than assets → negative net worth.
	assets := []valueRow{{value: 100, currency: "ZAR"}}
	liabs := []valueRow{{value: 200, currency: "ZAR"}}
	snap := simulateNetWorth(assets, nil, liabs, nil, "ZAR")

	if snap.NetWorth != -100 {
		t.Errorf("net worth = %.2f, want -100", snap.NetWorth)
	}
}

// ─── Budget progress edges ────────────────────────────────────────────────────

type budgetProgressCase struct {
	name      string
	budgeted  float64
	actual    float64
	wantOver  bool
	wantUnder bool
	wantZero  bool
}

func computeBudgetRemaining(budgeted, actual float64) float64 {
	return budgeted - actual
}

func TestBudgetProgressEdges(t *testing.T) {
	cases := []budgetProgressCase{
		{name: "under budget", budgeted: 1000, actual: 750, wantUnder: true},
		{name: "over budget", budgeted: 500, actual: 600, wantOver: true},
		{name: "exactly on budget", budgeted: 300, actual: 300},
		{name: "zero budget zero actual", budgeted: 0, actual: 0, wantZero: true},
		{name: "zero actual", budgeted: 400, actual: 0, wantUnder: true},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			remaining := computeBudgetRemaining(tc.budgeted, tc.actual)

			if tc.wantOver && remaining >= 0 {
				t.Errorf("expected over-budget (remaining < 0), got %.2f", remaining)
			}
			if tc.wantUnder && remaining <= 0 {
				t.Errorf("expected under-budget (remaining > 0), got %.2f", remaining)
			}
			if tc.wantZero && remaining != 0 {
				t.Errorf("expected zero remaining, got %.2f", remaining)
			}
			// "exactly on budget": remaining == 0
			if !tc.wantOver && !tc.wantUnder && !tc.wantZero {
				if remaining != 0 {
					t.Errorf("expected 0 remaining, got %.2f", remaining)
				}
			}
		})
	}
}

// ─── Goal progress ────────────────────────────────────────────────────────────

func TestGoalProgressComputation(t *testing.T) {
	cases := []struct {
		name          string
		target        float64
		current       float64
		wantPct       float64
		wantPctCapped float64 // after clamping to [0,100]
	}{
		{name: "half done", target: 1000, current: 500, wantPct: 50},
		{name: "fully achieved", target: 1000, current: 1000, wantPct: 100},
		{name: "over-contributed", target: 1000, current: 1100, wantPct: 100}, // clamped
		{name: "nothing saved", target: 1000, current: 0, wantPct: 0},
		{name: "zero target", target: 0, current: 500, wantPct: 0}, // avoid div/0
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			g := &Goal{TargetAmount: tc.target, CurrentAmount: tc.current}
			g.computeProgress()

			got := g.ProgressPct
			if got != tc.wantPct {
				t.Errorf("progress = %.2f, want %.2f", got, tc.wantPct)
			}
			// ProgressPct must always be in [0, 100].
			if got < 0 || got > 100 {
				t.Errorf("progress %.2f is outside [0, 100]", got)
			}
		})
	}
}

// ─── roundTwo ─────────────────────────────────────────────────────────────────

func TestRoundTwo(t *testing.T) {
	cases := []struct {
		input float64
		want  float64
	}{
		{33.33333, 33.33},
		{66.66666, 66.67},
		{100.0, 100.0},
		{0.0, 0.0},
		{-5.555, -5.56},
	}
	for _, tc := range cases {
		got := roundTwo(tc.input)
		if math.Abs(got-tc.want) > 0.005 {
			t.Errorf("roundTwo(%.5f) = %.5f, want %.5f", tc.input, got, tc.want)
		}
	}
}

// ─── validateBudgetInput ─────────────────────────────────────────────────────

func TestValidateBudgetInput(t *testing.T) {
	good := CreateBudgetInput{
		Name:      "Monthly",
		Period:    "monthly",
		StartDate: time.Now(),
		Currency:  "ZAR",
	}
	if err := validateBudgetInput(good); err != nil {
		t.Errorf("valid input rejected: %v", err)
	}

	// Empty name.
	bad := good
	bad.Name = ""
	if err := validateBudgetInput(bad); err == nil {
		t.Error("empty name should be rejected")
	}

	// Invalid period.
	bad = good
	bad.Period = "daily"
	if err := validateBudgetInput(bad); err == nil {
		t.Error("invalid period should be rejected")
	}

	// Missing currency.
	bad = good
	bad.Currency = ""
	if err := validateBudgetInput(bad); err == nil {
		t.Error("empty currency should be rejected")
	}
}

// ─── validateGoalInput ───────────────────────────────────────────────────────

func TestValidateGoalInput(t *testing.T) {
	good := CreateGoalInput{
		Name:         "Emergency Fund",
		Kind:         "savings",
		TargetAmount: 50_000,
		Currency:     "ZAR",
	}
	if err := validateGoalInput(good); err != nil {
		t.Errorf("valid goal rejected: %v", err)
	}

	bad := good
	bad.Name = ""
	if err := validateGoalInput(bad); err == nil {
		t.Error("empty name should be rejected")
	}

	bad = good
	bad.Kind = "lottery"
	if err := validateGoalInput(bad); err == nil {
		t.Error("invalid kind should be rejected")
	}

	bad = good
	bad.TargetAmount = 0
	if err := validateGoalInput(bad); err == nil {
		t.Error("zero target_amount should be rejected")
	}

	bad = good
	bad.TargetAmount = -1
	if err := validateGoalInput(bad); err == nil {
		t.Error("negative target_amount should be rejected")
	}
}

// ─── helpers ─────────────────────────────────────────────────────────────────

func nullString(s string) sql.NullString {
	return sql.NullString{String: s, Valid: s != ""}
}

File: worker/test/finance.test.ts
/**
 * Finance module unit tests.
 *
 * Tests cover the pure-logic parts of the port:
 *   1. Spending share % computation (decimal.js, no float drift).
 *   2. Budget-line remaining = amount − actual (decimal.js subtraction).
 *   3. Net-worth aggregation with FX conversion (divide by rate).
 *   4. Goal progress clamping (0–100).
 *   5. Serialisation helpers (roundTwo, money).
 *
 * These tests do NOT hit a real database; they exercise the math directly
 * using lib/money functions and inline re-implementations of the same logic
 * used in queries.ts / routes.ts, giving confidence that the port is exact.
 */
import { test, expect, describe } from "vitest";
import { dec, money, add, sub, sum } from "../src/lib/money";
import Decimal from "decimal.js";

// ─── Spending share % ─────────────────────────────────────────────────────────

describe("spendingBreakdown share percent", () => {
  /**
   * Inline port of the share calculation in queries.ts spendingBreakdown().
   * Input: NUMERIC amounts as strings (as returned by Neon).
   */
  function computeShares(
    rows: { total_amount: string }[],
  ): { share_pct: Decimal }[] {
    const grandTotal = sum(rows.map((r) => r.total_amount));
    return rows.map((r) => {
      const amt = dec(r.total_amount);
      const pct = grandTotal.isZero()
        ? new Decimal(0)
        : amt.div(grandTotal).times(100);
      return { share_pct: pct };
    });
  }

  test("two-category split sums to exactly 100%", () => {
    const rows = [
      { total_amount: "75.00" },
      { total_amount: "25.00" },
    ];
    const shares = computeShares(rows);
    const total = shares[0].share_pct.plus(shares[1].share_pct);
    expect(total.toFixed(2)).toBe("100.00");
  });

  test("three-category split — no float drift on 1/3 fractions", () => {
    const rows = [
      { total_amount: "100.00" },
      { total_amount: "100.00" },
      { total_amount: "100.00" },
    ];
    const shares = computeShares(rows);
    // Each should be 33.3333...%  Sum should be exactly 100.
    const total = shares.reduce(
      (acc, s) => acc.plus(s.share_pct),
      new Decimal(0),
    );
    // Can't be exactly 100 with 1/3 fractions, but it should be very close.
    // What matters is that it's NOT distorted by float64 rounding:
    expect(
      total.minus(100).abs().lessThan(new Decimal("0.0001")),
    ).toBe(true);
  });

  test("zero grand total yields 0% for all categories", () => {
    const rows = [{ total_amount: "0.00" }, { total_amount: "0.00" }];
    const shares = computeShares(rows);
    for (const s of shares) {
      expect(s.share_pct.isZero()).toBe(true);
    }
  });

  test("single category gets 100%", () => {
    const rows = [{ total_amount: "1234.56" }];
    const shares = computeShares(rows);
    expect(shares[0].share_pct.toFixed(2)).toBe("100.00");
  });

  test("no float drift on 0.1 + 0.2 total", () => {
    // Classic float trap: 0.1 + 0.2 ≠ 0.3 in float64.
    const rows = [
      { total_amount: "0.10" },
      { total_amount: "0.20" },
    ];
    const grandTotal = sum(rows.map((r) => r.total_amount));
    // Decimal arithmetic: 0.10 + 0.20 = 0.30 exactly.
    expect(grandTotal.toFixed(2)).toBe("0.30");

    const shares = computeShares(rows);
    // 0.10 / 0.30 * 100 = 33.333...
    // 0.20 / 0.30 * 100 = 66.666...
    expect(parseFloat(shares[0].share_pct.toFixed(2))).toBeCloseTo(33.33, 1);
    expect(parseFloat(shares[1].share_pct.toFixed(2))).toBeCloseTo(66.67, 1);
  });
});

// ─── Budget progress remaining ────────────────────────────────────────────────

describe("budgetProgress remaining", () => {
  /**
   * Inline port of the remaining = amount − actual calculation.
   * Inputs arrive as NUMERIC strings.
   */
  function computeRemaining(amount: string, actual: string): string {
    return sub(dec(amount), dec(actual)).toFixed(2);
  }

  test("on-budget: remaining = amount - actual", () => {
    expect(computeRemaining("1000.00", "650.00")).toBe("350.00");
  });

  test("over-budget: remaining is negative", () => {
    expect(computeRemaining("500.00", "750.00")).toBe("-250.00");
  });

  test("zero actual: remaining equals amount", () => {
    expect(computeRemaining("200.00", "0.00")).toBe("200.00");
  });

  test("exactly spent: remaining is zero", () => {
    expect(computeRemaining("123.45", "123.45")).toBe("0.00");
  });

  test("no float drift on fractional amounts", () => {
    // In float64: 100.10 - 0.10 might give 99.99999999 due to rounding.
    expect(computeRemaining("100.10", "0.10")).toBe("100.00");
  });
});

// ─── Net-worth FX conversion ──────────────────────────────────────────────────

describe("net-worth FX conversion", () => {
  /**
   * Port of convertToBase from queries.ts.
   * rate = 1 baseCurrency buys `rate` quoteCurrency.
   * Conversion: amountInBase = amountInQuote / rate.
   */
  function convertToBase(
    amount: string,
    currency: string,
    baseCurrency: string,
    fxRates: Map<string, string>,
  ): Decimal {
    const amt = dec(amount);
    if (currency === baseCurrency || !currency) return amt;
    const rateStr = fxRates.get(currency);
    if (rateStr) {
      const rate = dec(rateStr);
      if (!rate.isZero()) return amt.div(rate);
    }
    return amt; // fallback: treat as already in base
  }

  const fxRates = new Map([
    ["USD", "18.50"], // 1 ZAR = 18.50 USD? No — 1 ZAR buys 0.054 USD.
    // Correct interpretation: base=ZAR, quote=USD, rate=18.50 means
    // 1 ZAR = 18.50/18.50 ... Let's use the actual Go semantic:
    // rate is stored as: 1 base = rate quote.
    // So base=ZAR, quote=USD, rate=0.054 → 1 ZAR = 0.054 USD
    // → to convert 100 USD to ZAR: 100 / 0.054 ≈ 1851.85 ZAR
    // Let's use realistic rates for testing:
    ["EUR", "20.00"], // 1 ZAR = 0.05 EUR → rate = 0.05
    // But rate stored in DB is base=ZAR, quote=EUR, so 1 ZAR = rate EUR.
    // Let's be concrete: ZAR base, USD quote, rate = 0.054
  ]);

  // Use a cleaner example: base=ZAR, USD rate = 18 (i.e. 1 USD = 18 ZAR,
  // so rate = 18 if stored as 1 ZAR = 1/18 USD ... no.
  //
  // Go semantics: fx_rates.base = "ZAR", fx_rates.quote = "USD",
  // fx_rates.rate = 18.5  means 1 ZAR = 18.5 USD? That would be strange.
  // Let's re-read the Go code:
  //   "rate is quote/base, so divide to get base"
  //   convert: amount / rate  where rate = quotePerBase
  // So if base=ZAR, quote=USD, rate = 0.054:
  //   100 USD / 0.054 = 1851.85 ZAR ✓
  //
  // Alternatively if base=ZAR, quote=USD, rate = 18.5:
  //   100 USD / 18.5 = 5.4 ZAR ✗
  //
  // Correct: rate = how many quote per 1 base.
  // base=ZAR, quote=USD → if 1 ZAR = 0.054 USD then rate = 0.054
  // 100 USD to ZAR = 100 / 0.054

  const zarRates = new Map([
    ["USD", "0.054"], // 1 ZAR = 0.054 USD
    ["GBP", "0.043"], // 1 ZAR = 0.043 GBP
  ]);

  test("same currency returns amount unchanged", () => {
    const result = convertToBase("1000.00", "ZAR", "ZAR", zarRates);
    expect(result.toFixed(2)).toBe("1000.00");
  });

  test("converts USD asset to ZAR base (divide by rate)", () => {
    // 540 USD / 0.054 = 10000 ZAR
    const result = convertToBase("540.00", "USD", "ZAR", zarRates);
    expect(result.toFixed(2)).toBe("10000.00");
  });

  test("converts GBP liability to ZAR base", () => {
    // 430 GBP / 0.043 = 10000 ZAR
    const result = convertToBase("430.00", "GBP", "ZAR", zarRates);
    expect(result.toFixed(2)).toBe("10000.00");
  });

  test("unknown currency passes through unchanged (best-effort)", () => {
    const result = convertToBase("999.00", "CHF", "ZAR", zarRates);
    expect(result.toFixed(2)).toBe("999.00");
  });

  test("net-worth aggregation: assets + holdings - liabilities", () => {
    const rates = new Map([["USD", "0.05"]]); // 1 ZAR = 0.05 USD

    const assets = [
      { value: "500.00", currency: "ZAR" },  // 500 ZAR
      { value: "50.00", currency: "USD" },   // 50/0.05 = 1000 ZAR
    ];
    const holdings = [
      { value: "200.00", currency: "ZAR" },  // 200 ZAR
    ];
    const liabs = [
      { value: "300.00", currency: "ZAR" },  // 300 ZAR
    ];

    const totalAssets = sum(
      assets.map((a) => convertToBase(a.value, a.currency, "ZAR", rates)),
    );
    const totalHoldings = sum(
      holdings.map((h) => convertToBase(h.value, h.currency, "ZAR", rates)),
    );
    const totalLiabs = sum(
      liabs.map((l) => convertToBase(l.value, l.currency, "ZAR", rates)),
    );

    // totalAssets = 500 + 1000 = 1500
    expect(totalAssets.toFixed(2)).toBe("1500.00");
    // totalHoldings = 200
    expect(totalHoldings.toFixed(2)).toBe("200.00");
    // totalLiabs = 300
    expect(totalLiabs.toFixed(2)).toBe("300.00");

    // net = 1500 + 200 - 300 = 1400
    const netWorth = add(add(totalAssets, totalHoldings), totalLiabs.neg());
    expect(money(netWorth)).toBe("1400.00");
  });

  test("FX division is exact (no float drift on 100 / 0.05)", () => {
    // float64: 100 / 0.05 can introduce rounding error.
    const result = convertToBase("100.00", "USD", "ZAR", new Map([["USD", "0.05"]]));
    expect(result.toFixed(2)).toBe("2000.00");
  });
});

// ─── Goal progress ────────────────────────────────────────────────────────────

describe("goal progress percent", () => {
  /** Port of Go goal.computeProgress — uses float here since Go does too. */
  function computeProgress(targetAmount: string, currentAmount: string): number {
    const target = parseFloat(money(targetAmount));
    const current = parseFloat(money(currentAmount));
    if (target <= 0) return 0;
    const pct = (current / target) * 100;
    return Math.min(100, Math.max(0, pct));
  }

  /** roundTwo port (Go: roundTwo). */
  function roundTwo(n: number): number {
    return parseFloat(new Decimal(n).toFixed(2));
  }

  test("50% progress", () => {
    expect(roundTwo(computeProgress("1000.00", "500.00"))).toBe(50);
  });

  test("100% when fully achieved", () => {
    expect(roundTwo(computeProgress("500.00", "500.00"))).toBe(100);
  });

  test("clamped to 100 when over-achieved", () => {
    expect(roundTwo(computeProgress("500.00", "600.00"))).toBe(100);
  });

  test("0 when no progress", () => {
    expect(roundTwo(computeProgress("1000.00", "0.00"))).toBe(0);
  });

  test("0 when target is zero (guard)", () => {
    expect(roundTwo(computeProgress("0.00", "100.00"))).toBe(0);
  });

  test("partial progress rounds to 2dp", () => {
    // 333.33 / 1000 * 100 = 33.333% → rounded to 33.33
    const pct = computeProgress("1000.00", "333.33");
    expect(roundTwo(pct)).toBe(33.33);
  });
});

// ─── money helper invariants ──────────────────────────────────────────────────

describe("money helper invariants", () => {
  test("add is exact (0.1 + 0.2 = 0.30)", () => {
    expect(money(add("0.1", "0.2"))).toBe("0.30");
  });

  test("sub is exact (100.10 - 0.10 = 100.00)", () => {
    expect(money(sub("100.10", "0.10"))).toBe("100.00");
  });

  test("sum of empty array is 0", () => {
    expect(sum([]).toFixed(2)).toBe("0.00");
  });

  test("dec handles null/undefined gracefully (returns 0)", () => {
    expect(dec(null).toFixed(2)).toBe("0.00");
    expect(dec(undefined).toFixed(2)).toBe("0.00");
    expect(dec("").toFixed(2)).toBe("0.00");
  });
});
